### PR TITLE
Add Ellipsis typehint to reductions

### PR DESCRIPTION
**Pull request by headtr1ck**
This PR adds the ellipsis typehint to reductions (only where they behave differently from None to reduce overhead).
Follow up on https://github.com/pydata/xarray/pull/7017#issuecomment-1243927061

Additionally I was changing a lot of "one or more dimensions" typehints to `str | Iterable[Hashable]` (See https://github.com/pydata/xarray/issues/6142).
Some code changes were necessary to support this fully. Before several things were not working with actual hashable dimensions that are not strings.

### CODE REVIEW

**Comment by max-sixty** [comment]
Excellent @headtr1ck !

Do we need to run `pytest --accept` to get the docstrings? It looks like we lost lots...

**Comment by headtr1ck** [comment]
Could any dev that uses linux rerun the generate_reductions and `pytest --doctest-modules xarray/core/_reductions.py --accept`?
On windows I still get different results (maybe that should be fixed at some point...)

File: xarray/util/generate_reductions.py [hunk 1]
```diff
@@ -45,9 +46,9 @@ class {obj}{cls}Reductions:
     def reduce(
         self,
         func: Callable[..., Any],
-        dim: None | Hashable | Sequence[Hashable] = None,
+        dim: str | Iterable[Hashable] | None = None,
```
Comment (Illviljan) [inline xarray/util/generate_reductions.py:49]:
```suggestion
        dim: T_Dim = None,
```
Should this become a variable? It should make it easier adding the typing in `flox` easier as well.

**Comment by headtr1ck** [comment]
Turns out that the buildin `ellipsis` works now with mypy.
Did not test it for older python versions, may require some special casing (Lets see if the tests pass)?

**Comment by max-sixty** [comment]
Here's the diff from pytest-accept (it is weird that it's slightly different on windows...)

```diff
commit 83615a94a6b7c0ae0cf0e0240d7705d9ce6c21e5
Author: Maximilian Roos <m@maxroos.com>
Date:   Sat Sep 17 12:36:45 2022 -0700

    pytest-accept

diff --git a/xarray/core/_reductions.py b/xarray/core/_reductions.py
index a7cf7ec2..d0c2a9d7 100644
--- a/xarray/core/_reductions.py
+++ b/xarray/core/_reductions.py
@@ -97,7 +97,7 @@ def count(
         <xarray.Dataset>
         Dimensions:  ()
         Data variables:
-            da       int32 5
+            da       int64 5
         """
         return self.reduce(
             duck_array_ops.count,
@@ -4400,7 +4400,7 @@ def count(
 
         >>> da.groupby("labels").count()
         <xarray.DataArray (labels: 3)>
-        array([1, 2, 2], dtype=int64)
+        array([1, 2, 2])
         Coordinates:
           * labels   (labels) object 'a' 'b' 'c'
         """
@@ -5485,7 +5485,7 @@ def count(
 
         >>> da.resample(time="3M").count()
         <xarray.DataArray (time: 3)>
-        array([1, 3, 1], dtype=int64)
+        array([1, 3, 1])
         Coordinates:
           * time     (time) datetime64[ns] 2001-01-31 2001-04-30 2001-07-31
         """
```

**Review by Illviljan — COMMENTED** [review]
I wonder if Dims should include `ellipsis` as well? The few times it's missing might be issues with the functions?

**Comment by Illviljan** [comment]
> ```diff
> @@ -4400,7 +4400,7 @@ def count(
>  
>          >>> da.groupby("labels").count()
>          <xarray.DataArray (labels: 3)>
> -        array([1, 2, 2], dtype=int64)
> +        array([1, 2, 2])
>          Coordinates:
>            * labels   (labels) object 'a' 'b' 'c'
>          """
> ```

Is it just me that this example crashes the second time I run it?
``` python
import numpy as np
import pandas as pd
import xarray as xr

da = xr.DataArray(
    np.array([1, 2, 3, 1, 2, np.nan]),
    dims="time",
    coords=dict(
        time=("time", pd.date_range("01-01-2001", freq="M", periods=6)),
        labels=("time", np.array(["a", "b", "c", "c", "b", "a"])),
    ),
)
da.groupby("labels").count()
```

**Comment by headtr1ck** [comment]
> Is it just me that this example crashes the second time I run it?

Could you specify what you mean by "second time I run it"?
Executing the groupby twice?

**Comment by Illviljan** [comment]
Just running that script file several times without restarting the console. It might be a Spyder bug though since I can't reproduce it in a stand alone ipython console.

This for example works (the first time):
``` python
import numpy as np
import pandas as pd
import xarray as xr

da = xr.DataArray(
    np.array([1, 2, 3, 1, 2, np.nan]),
    dims="time",
    coords=dict(
        time=("time", pd.date_range("01-01-2001", freq="M", periods=6)),
        labels=("time", np.array(["a", "b", "c", "c", "b", "a"])),
    ),
)
da.groupby("labels").count()

da = xr.DataArray(
    np.array([1, 2, 3, 1, 2, np.nan]),
    dims="time",
    coords=dict(
        time=("time", pd.date_range("01-01-2001", freq="M", periods=6)),
        labels=("time", np.array(["a", "b", "c", "c", "b", "a"])),
    ),
)
da.groupby("labels").count()
```

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -5065,15 +5063,15 @@ def drop_isel(self: T_Dataset, indexers=None, **indexers_kwargs) -> T_Dataset:
 
     def drop_dims(
         self: T_Dataset,
-        drop_dims: Hashable | Iterable[Hashable],
+        drop_dims: str | Iterable[Hashable],
```
Comment (Illviljan) [inline xarray/core/dataset.py:5066]: If `Dims = Union[str, Iterable[Hashable]]` this one can be replaced as well.
I think the other ones will still be ok as  `dim: Dims = None` will implicitly add None.
```python
def test(a: str = None):
    reveal_type(a)  # note: Revealed type is "Union[builtins.str, None]"
```

Comment (headtr1ck) [reply xarray/core/dataset.py:5066]: I think explicit is always better, especially when switching to sphinx to write the docstring types.

Comment (Illviljan) [reply xarray/core/dataset.py:5066]:
```suggestion
        drop_dims: Dims = None
```
I think aligning all these dim inputs with a single Dims type is a bigger win. But if we're going the explicit route then maybe this function should have None as default?

Comment (headtr1ck) [reply xarray/core/dataset.py:5066]: Should I do it in this PR still?

Comment (Illviljan) [reply xarray/core/dataset.py:5066]: It can be done in a different PR if you prefer that. This is already an improvement.

**Comment by headtr1ck** [comment]
> ```python
> import numpy as np
> import pandas as pd
> import xarray as xr
> 
> da = xr.DataArray(
>     np.array([1, 2, 3, 1, 2, np.nan]),
>     dims="time",
>     coords=dict(
>         time=("time", pd.date_range("01-01-2001", freq="M", periods=6)),
>         labels=("time", np.array(["a", "b", "c", "c", "b", "a"])),
>     ),
> )
> da.groupby("labels").count()
> 
> da = xr.DataArray(
>     np.array([1, 2, 3, 1, 2, np.nan]),
>     dims="time",
>     coords=dict(
>         time=("time", pd.date_range("01-01-2001", freq="M", periods=6)),
>         labels=("time", np.array(["a", "b", "c", "c", "b", "a"])),
>     ),
> )
> da.groupby("labels").count()
> ```

For me this works in a python terminal, python script and jupyter notebook (I don't use spyder but vscode).

File: xarray/core/dataset.py (hunk 2 above)
Comment (headtr1ck) [reply xarray/core/dataset.py:5066]: Ok, then lets do it in a different PR :) I will  not forget about it.

**Pull request merged**